plugins/movetoarchive.koplugin/main.lua

```diff
@@ -0,0 +1,161 @@
+                        self:showNoArchiveConfirmBox()
+                        return
+                    end
+                    self:openFileBrowser(self.archive_dir_path)
+                end,
+            },
+            {
+                text = _("Go to last copied/moved from folder"),
+                callback = function()
+                    if not self.last_copied_from_dir then
+                        UIManager:show(InfoMessage:new{
+                            text = _("No previous folder found.")
+                         })
+                        return
+                    end
+                    self:openFileBrowser(self.last_copied_from_dir)
+                end,
+            },
+            {
+                text = _("Set archive directory"),
+                keep_menu_open = true,
+                callback =  self.setArchiveDirectory,
+            }
+        },
+    }
+end
+
+function MoveToArchive:moveToArchive()
+    local move_done_text = _("Book moved.\nDo you want to open it from the archive folder?")
+    self:commonProcess(true, move_done_text)
+end
+
+function MoveToArchive:copyToArchive()
+    local copy_done_text =_("Book copied.\nDo you want to open it from the archive folder?")
+    self:commonProcess(false, copy_done_text)
+end
+
+function MoveToArchive:commonProcess(is_move_process, moved_done_text)
+    if not self.archive_dir_path then
+        self:showNoArchiveConfirmBox()
+        return
+    end
+    local document_full_path = self.ui.document.file
+    local filename
+    self.last_copied_from_dir, filename = util.splitFilePathName(document_full_path)
+
+    logger.dbg("MoveToArchive: last_moved/copied_from_dir :", self.last_copied_from_dir)
+    self.move_to_archive_settings:saveSetting("last_copied_from_dir", ("%s/"):format(self.last_copied_from_dir))
+
+    self.ui:onClose()
+    if is_move_process then
+        FileManager:moveFile(document_full_path, self.archive_dir_path)
+        FileManager:moveFile(DocSettings:getSidecarDir(document_full_path), self.archive_dir_path)
+    else
+        FileManager:copyFileFromTo(document_full_path, self.archive_dir_path)
+        FileManager:copyRecursive(DocSettings:getSidecarDir(document_full_path), self.archive_dir_path)
+    end
+    local dest_file = string.format("%s/%s", self.archive_dir_path, filename)
+    ReadHistory:updateItemByPath(document_full_path, dest_file)
+    ReadCollection:updateItemByPath(document_full_path, dest_file)
+    -- Update last open file.
+    if G_reader_settings:readSetting("lastfile") == document_full_path then
+        G_reader_settings:saveSetting("lastfile", dest_file)
+    end
+
+    UIManager:show(ConfirmBox:new{
+    text = moved_done_text,
+        ok_callback = function ()
+            ReaderUI:showReader(self.archive_dir_path .. filename)
+        end,
+        cancel_callback = function ()
+            if DocSettings:open(dest_file).data.percent_finished == 1 then
```

Comment:
(We should use `:readSetting("percent_finished")` instead of accessing the internal `data` table, but I see other plugins do that :/)

Anyway, that might be out of that plugin contract, to remove an (archived) finished book from History.
I would personally don't care about having it removed and having a clean History, but @Frenzie might :) He likes to keep even deleted files in History - and you're not even deleting it here: you have moved it into the archive directory.
Unless it can feel like implied by "Cancel". Or you should ask the user with another ConfirmBox?